File: src/main/java/ee/leola/kassa/controllers/rest/TransactionController.java
package ee.leola.kassa.controllers.rest;

import ee.leola.kassa.models.Product;
import ee.leola.kassa.models.Transaction;
import ee.leola.kassa.models.User;
import ee.leola.kassa.repository.ProductRepository;
import ee.leola.kassa.repository.TransactionRepository;
import ee.leola.kassa.repository.UserRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;

@RestController
@RequestMapping(value = "/transaction")
@Slf4j
public class TransactionController extends RestBaseController<Transaction, Long> {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private ProductRepository productRepository;

    @Autowired
    public TransactionController(TransactionRepository transactionRepository) {
        super(transactionRepository);
    }

    @RequestMapping(value = "/invalid/{id}", method = RequestMethod.POST)
    public void invalidate(@PathVariable("id") Long id) {
        invalidateTransaction(id);
    }

    @RequestMapping(value = "/invalid/admin/{id}", method = RequestMethod.POST)
    public void invalidateAdmin(@PathVariable("id") Long id) {
        invalidateTransaction(id);
    }

    private void invalidateTransaction(Long id) {
        Transaction transaction = repo.findById(id).orElseThrow(IllegalArgumentException::new);

        //Increment product quantities
        if (transaction.getPaytype().isAffectsQuantity()) {
            transaction.getProducts().forEach(product -> incrementProductQuantity(product.getProduct().getId(), product.getQuantity()));
        }

        if (transaction.getPaytype().isAffectsBalance()) {
            BigDecimal balance = transaction.getProducts().stream().map((product -> product.getPrice().multiply(BigDecimal.valueOf(product.getQuantity()))))
                    .reduce(BigDecimal.ZERO, BigDecimal::add);
            //Increment user balance
            changeUserBalance(transaction.getUser().getId(), balance);
        }

        transaction.setInvalid(true);
        repo.save(transaction);
    }

    @Override
    @RequestMapping(value = "/{id}", method = RequestMethod.POST)
    public ResponseEntity<Transaction> update(@PathVariable Long id, @RequestBody Transaction json) {
        return new ResponseEntity<>(HttpStatus.METHOD_NOT_ALLOWED);
    }

    @Override
    @RequestMapping(method = RequestMethod.POST)
    public Transaction create(@RequestBody Transaction transaction) {
        log.info("Saving new transaction {}", transaction);

        //Decrement product quantities and save
        if (transaction.getPaytype().isAffectsQuantity()) {
            transaction.getProducts().stream().forEach(product -> incrementProductQuantity(product.getProduct().getId(), -product.getQuantity()));
        }

        if (transaction.getPaytype().isAffectsBalance()) {
            BigDecimal balance = transaction.getProducts().stream().map((product -> product.getPrice().multiply(BigDecimal.valueOf(product.getQuantity()))))
                    .reduce(BigDecimal.ZERO, BigDecimal::subtract);
            //Decrement user balance
            changeUserBalance(transaction.getUser().getId(), balance);
        }


        repo.save(transaction);

        return transaction;
    }

    private void incrementProductQuantity(Long id, Integer quantity) {
        Product product = productRepository.findById(id).orElseThrow(IllegalArgumentException::new);
        product.incrementQuantity(quantity);
        productRepository.save(product);
    }

    private void changeUserBalance(Long id, BigDecimal balance) {
        User user = userRepository.findById(id).orElseThrow(IllegalArgumentException::new);
        user.setBalance(user.getBalance().add(balance));
        userRepository.save(user);
    }
}

File: src/main/kotlin/ee/leola/kassa/controllers/rest/TransactionController.kt
package ee.leola.kassa.controllers.rest

import ee.leola.kassa.helpers.LoggerDelegate
import ee.leola.kassa.models.Transaction
import ee.leola.kassa.repository.ProductRepository
import ee.leola.kassa.repository.TransactionRepository
import ee.leola.kassa.repository.UserRepository
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal

@RestController
@RequestMapping("/transaction")
class TransactionController(
    transactionRepository: TransactionRepository,
    private val userRepository: UserRepository,
    private val productRepository: ProductRepository,
) : RestBaseController<Transaction, Long>(transactionRepository) {
    private val logger by LoggerDelegate()

    @PostMapping("/invalid/{id}")
    fun invalidate(
        @PathVariable("id") id: Long,
    ) = invalidateTransaction(id)

    @PostMapping("/invalid/admin/{id}")
    fun invalidateAdmin(
        @PathVariable("id") id: Long,
    ) = invalidateTransaction(id)

    private fun invalidateTransaction(id: Long) {
        val transaction = repo.findById(id).orElseThrow()

        // Increment product quantities
        if (transaction.affectsQuantity()) {
            transaction.products.forEach { product ->
                incrementProductQuantity(product.product.id!!, product.quantity)
            }
        }
        if (transaction.affectsBalance()) {
            val balance = transaction.products.sumOf { it.price * it.quantity.toBigDecimal() }
            // Increment user balance
            changeUserBalance(transaction.user.id!!, balance)
        }
        repo.save(transaction.invalidate())
    }

    @PostMapping("/{id}")
    override fun update(
        @PathVariable id: Long,
        @RequestBody json: Transaction,
    ): ResponseEntity<Transaction> {
        return ResponseEntity(HttpStatus.METHOD_NOT_ALLOWED)
    }

    @PostMapping
    override fun create(
        @RequestBody json: Transaction,
    ): Transaction {
        logger.info("Saving new transaction {}", json)
        // Decrement product quantities and save
        if (json.affectsQuantity()) {
            json.products.forEach { product ->
                incrementProductQuantity(product.product.id!!, -product.quantity)
            }
        }
        if (json.affectsBalance()) {
            val balance = json.products.sumOf { it.price * it.quantity.toBigDecimal() }
            // Decrement user balance
            changeUserBalance(json.user.id!!, -balance)
        }
        repo.save(json)
        return json
    }

    private fun incrementProductQuantity(
        id: Long,
        quantity: Int,
    ) {
        val product = productRepository.findById(id).orElseThrow()
        productRepository.save(product.incrementQuantity(quantity))
    }

    private fun changeUserBalance(
        id: Long,
        balance: BigDecimal,
    ) {
        val user = userRepository.findById(id).orElseThrow()
        userRepository.save(user.incrementBalance(balance))
    }
}
